addons/binding/org.openhab.binding.solaredge/src/main/java/org/openhab/binding/solaredge/handler/SolarEdgeAggregateDataPolling.java

```diff
@@ -0,0 +1,88 @@
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.solaredge.handler;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.openhab.binding.solaredge.internal.command.AggregateDataUpdatePrivateApi;
+import org.openhab.binding.solaredge.internal.command.AggregateDataUpdatePublicApi;
+import org.openhab.binding.solaredge.internal.command.SolarEdgeCommand;
+import org.openhab.binding.solaredge.internal.model.AggregatePeriod;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Polling worker class. This is responsible for periodic polling of sensor data.
+ *
+ * @author Alexander Friese - initial contribution
+ */
+public class SolarEdgeAggregateDataPolling implements Runnable {
+    /**
+     * Logger
+     */
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    /**
+     * Handler for delegation to callbacks.
+     */
+    private final SolarEdgeHandler handler;
+
+    /**
+     * Constructor.
+     *
+     * @param handler handler which handles results of polling
+     */
+    public SolarEdgeAggregateDataPolling(SolarEdgeHandler handler) {
+        this.handler = handler;
+    }
+
+    /**
+     * Poll the SolarEdge Webservice one time per call.
+     */
+    @Override
+    public void run() {
+        if (handler.getWebInterface() != null) {
+            // if no meter is present all data will be fetched by the 'LiveDataUpdateMeterless'
+            if (handler.getConfiguration().isMeterInstalled()) {
+                logger.debug("polling SolarEdge aggregate data {}", handler.getConfiguration());
+
+                List<SolarEdgeCommand> commands = new ArrayList<>();
+
+                if (handler.getConfiguration().isUsePrivateApi()) {
+                    commands.add(new AggregateDataUpdatePrivateApi(handler, AggregatePeriod.DAY));
+                    commands.add(new AggregateDataUpdatePrivateApi(handler, AggregatePeriod.WEEK));
+                    commands.add(new AggregateDataUpdatePrivateApi(handler, AggregatePeriod.MONTH));
+                    commands.add(new AggregateDataUpdatePrivateApi(handler, AggregatePeriod.YEAR));
+                } else {
+                    commands.add(new AggregateDataUpdatePublicApi(handler, AggregatePeriod.DAY));
+                    commands.add(new AggregateDataUpdatePublicApi(handler, AggregatePeriod.WEEK));
+                    commands.add(new AggregateDataUpdatePublicApi(handler, AggregatePeriod.MONTH));
+                    commands.add(new AggregateDataUpdatePublicApi(handler, AggregatePeriod.YEAR));
+                }
+
+                try {
+                    for (SolarEdgeCommand command : commands) {
+                        handler.getWebInterface().executeCommand(command);
+                        // we must not flood the API with requests, otherwise we will get an exception
+                        Thread.sleep(5000);
```

Comment:
You are blocking a thread of the thread pool here for most of its time - this is pretty bad.
You should instead treat every command as a separate task that you put in a queue and and then have a job that is scheduled every 5 seconds and which picks up and executes the first command of the queue (if there is any).

Reply:
This is now solved